stl/inc/chrono

```diff
@@ -652,6 +687,1103 @@ namespace chrono {
+            const int _Z        = _Tp + 719468;
+            const int _Era      = (_Z >= 0 ? _Z : _Z - 146096) / 146097;
+            const unsigned _Doe = static_cast<unsigned>(_Z - _Era * 146097);
+            const unsigned _Yoe = (_Doe - _Doe / 1460 + _Doe / 36524 - _Doe / 146096) / 365;
+            const int _Y        = static_cast<int>(_Yoe) + _Era * 400;
+            const unsigned _Doy = _Doe - (365 * _Yoe + _Yoe / 4 - _Yoe / 100);
+            const unsigned _Mp  = (5 * _Doy + 2) / 153;
+            const unsigned _D   = _Doy - (153 * _Mp + 2) / 5 + 1;
+            const unsigned _M   = _Mp + (_Mp < 10 ? 3 : static_cast<unsigned>(-9));
+            return year_month_day(chrono::year{_Y + (_M <= 2)}, chrono::month{_M}, chrono::day{_D});
+        }
+        _NODISCARD constexpr days _Days_from_civil() const noexcept {
+            static_assert(_STD numeric_limits<unsigned>::digits >= 18);
+            static_assert(_STD numeric_limits<int>::digits >= 20);
+            const int _Y        = static_cast<int>(_Year) - (_Month <= chrono::month{2});
+            const unsigned _M   = static_cast<unsigned>(_Month);
+            const int _Era      = (_Y >= 0 ? _Y : _Y - 399) / 400;
+            const unsigned _Yoe = static_cast<unsigned>(_Y - _Era * 400);
+            const unsigned _Doy =
+                (153 * (_M + (_M > 2 ? static_cast<unsigned>(-3) : 9)) + 2) / 5 + static_cast<unsigned>(_Day) - 1;
+            const unsigned _Doe = _Yoe * 365 + _Yoe / 4 - _Yoe / 100 + _Doy;
+            return days{_Era * 146097 + static_cast<int>(_Doe) - 719468};
+        }
+
+    public:
+        year_month_day() = default;
+        constexpr year_month_day(const chrono::year& _Y, const chrono::month& _M, const chrono::day& _D) noexcept
+            : _Year{_Y}, _Month{_M}, _Day{_D} {}
+        constexpr year_month_day(const class year_month_day_last& _Ymdl) noexcept;
+        constexpr year_month_day(const sys_days& _Sys_days) noexcept
+            : year_month_day{_Civil_from_days(_Sys_days.time_since_epoch().count())} {}
+        constexpr explicit year_month_day(const local_days& _Local_days) noexcept
+            : year_month_day{_Civil_from_days(_Local_days.time_since_epoch().count())} {}
+
+        constexpr year_month_day& operator+=(const months& _Months) noexcept;
+        constexpr year_month_day& operator-=(const months& _Months) noexcept;
+        constexpr year_month_day& operator+=(const years& _Years) noexcept;
+        constexpr year_month_day& operator-=(const years& _Years) noexcept;
+
+        _NODISCARD constexpr chrono::year year() const noexcept {
+            return _Year;
+        }
+        _NODISCARD constexpr chrono::month month() const noexcept {
+            return _Month;
+        }
+        _NODISCARD constexpr chrono::day day() const noexcept {
+            return _Day;
+        }
+
+        _NODISCARD constexpr operator sys_days() const noexcept {
+            return sys_days{_Days_from_civil()};

```

Comment:
That should be what the algorithm does, no? The `!ok()` case is unspecified.

Reply:
The case I don't see covered is the 
> Otherwise, if y_­.ok() && m_­.ok() is true, returns sys_­days{y_­/m_­/1d} + (d_­ - 1d). 

The "otherwise" is referring to when `y_.ok() && m_.ok()` is true, but `d_` is not a valid day for that year/month.  Currently, `sys_days()` currently just calls `_Days_from_civil()` unconditionally and the "otherwise" case mentioned here is not covered in the precondition for that function that is listed in the date algorithms paper.  I haven't traced through the `_Days_from_civil()` function closely enough to see if it happens to be doing the right thing in this case, but the fact that it fails the preconditions makes me think that we should be handling this case specially and only calling `_Days_from_civil()` when `*this` is `ok()` and handling the other case specially.  

On that note, I think it would be good to document the preconditions for each of the functions from the date algorithms paper.

Reply:
It should be covered
https://github.com/microsoft/STL/blob/1fbc15132ff5f398b0733140999415fbad38a25c/tests/std/tests/P0355R7_calendars_and_time_zones_dates/test.cpp#L362

Reply:
I'm still a bit skeptical considering the documented preconditions.  We should either get clarification from Howard that the `days_from_civil` algorithm covers this case (when `y_.ok() && m_.ok()` is true, but `d_` is not a valid day for that year/month), or handle this case that is not covered in the algorithm's preconditions separately.  (This assessment supported by @StephanTLavavej.)

Reply:
You're ok as is.  `_Days_from_civil` uses `_Day` in only one place:
```
const unsigned _Doy =
   (153 * (_M + (_M > 2 ? static_cast<unsigned>(-3) : 9)) + 2) / 5 + static_cast<unsigned>(_Day) - 1;
```
And in that one place it is simply adding `_Day` (minus 1) to the otherwise accumulated count of days.  That is when `y_.ok() && m_.ok()` is true, then you're getting a correct count such that if `_Day == 1`, you get the correct count corresponding to the first day of the month.  If `_Day == 0`, you get the day before that.  If `_Day == 55`, you get 54 days after the first day of the month.  And all of this is desirable, and the intended specification, and your current implementation.